tests/unit/TypeInferenceTests.d

```diff
@@ -0,0 +1,217 @@
+/**
+ * Copyright: Copyright (c) 2017 Wojciech Szęszoł. All rights reserved.
+ * Authors: Wojciech Szęszoł
+ * Version: Initial created: October 13, 2017
+ * License: $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0)
+ */
+import core.exception;
+
+import std.container.dlist;
+import std.variant;
+
+import Common;
+import Assert;
+
+import dstep.translator.MacroDefinition;
+
+unittest
+{
+    assertTranslates(q"C
+#define FOO(x) BAR(int)
+C", q"D
+extern (C):
+
+extern (D) auto FOO(T)(auto ref T x)
+{
+    return BAR(int);
```

Comment:
This is translated like this because at this point we don't know what `BAR` is?

Reply:
Yes, do you think BAR!int() is better?

Reply:
Well, `BAR(int);` doesn't compile, regardless if `BAR` is defined in the D code or not. So yes I think `BAR!int()` is better.

Reply:
It turns out that it isn't that trivial to change it (more than 1h of work), can we merge it as is for now? I don't have time to fix it right now. Maybe I'll manage after winter...

Reply:
Sure, can you rebase please.

Reply:
Do you remember what did you mean by rebase? There is no merge conflict after all.

Reply:
No. The only thing I can think of is perhaps it's behind master.

Reply:
Rebased.